homeassistant/helpers/entity.py

```diff
@@ -1430,10 +1447,23 @@ async def async_internal_will_remove_from_hass(self) -> None:
         if self.platform:
             self.hass.data[DATA_ENTITY_SOURCE].pop(self.entity_id)
 
-    async def _async_registry_updated(
+    @callback
```

Comment:
We should run this run_immediately because otherwise it's a call soon and there is still a race

Reply:
This one doesn't support that unfortunately so this is the best we can do without a complete refactor of the event layer